src/Microsoft.AspNet.Server.Kestrel/Http/MemoryPoolBlock2.cs

```diff
@@ -0,0 +1,551 @@
+                            var ch0Count = Vector.Dot(ch0Equals, _dotCount);
+                            var ch1Equals = Vector.Equals(data, ch1Vector);
+                            var ch1Count = Vector.Dot(ch1Equals, _dotCount);
+
+                            if (ch0Count == 0 && ch1Count == 0)
+                            {
+                                index += vectorStride;
+                                continue;
+                            }
+                            else if (ch0Count < 2 && ch1Count < 2)
+                            {
+                                var ch0Index = ch0Count == 1 ? Vector.Dot(ch0Equals, _dotIndex) : byte.MaxValue;
+                                var ch1Index = ch1Count == 1 ? Vector.Dot(ch1Equals, _dotIndex) : byte.MaxValue;
+                                if (ch0Index < ch1Index)
+                                {
+                                    _block = block;
+                                    _index = index + ch0Index;
+                                    return char0;
+                                }
+                                else
+                                {
+                                    _block = block;
+                                    _index = index + ch1Index;
+                                    return char1;
+                                }
+                            }
+                            else
+                            {
+                                following = vectorStride;
+                            }
+                        }
+                        for (; following != 0; following--, index++)
+                        {
+                            var byteIndex = block.Array[index];
+                            if (byteIndex == byte0)
+                            {
+                                _block = block;
+                                _index = index;
+                                return char0;
+                            }
+                            else if (byteIndex == byte1)
+                            {
+                                _block = block;
+                                _index = index;
+                                return char1;
+                            }
+                        }
+                    }
+                }
+            }
+
+            public int GetLength(Iterator end)
+            {
+                if (IsDefault || end.IsDefault)
+                {
+                    return -1;
+                }
+
+                var block = _block;
+                var index = _index;
+                var length = 0;
+                while (true)
+                {
+                    if (block == end._block)
+                    {
+                        return length + end._index - index;
+                    }
+                    else if (block.Next == null)
+                    {
+                        throw new Exception("end did not follow iterator");
```

Comment:
Yes - now everything that happens inside the frame's asyc processing function (which includes framing, application invocation, and all calls back into stream read/write) will ultimately land at the by the processing function's catch/finally (if they're not caught by the app or error handling middleware).

The processing function's ultimate catch/finally will disconnect the socket, and everything will teardown.